### PR TITLE
Adding docker registry sync

### PR DESCRIPTION
**Pull request by GitHK**
# What do these changes contain

- [x] Adding a new service used for syncing images between different repositories. More details can be found in the project's `README.md` file.
- [x] A common defined way to merge build and push images to Docker Hub

### CODE REVIEW

File: ci/github/action-build-and-push/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+FROM jpetazzo/dind
+
+RUN apt-get update && apt-get install -y make curl git jq
```
Comment (pcrespov) [inline ci/github/action-build-and-push/Dockerfile:3]: MINOR: list every lib in a different line (see [best practices in Docker](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#sort-multi-line-arguments)), i.e

```Dockerfile
RUN apt-get update && apt-get install -y \
         make \
         curl \
         git \
         jq  
```

File: ci/github/action-build-and-push/action.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+name: "Build&Push to Docker Hub"
```
Comment (pcrespov) [inline ci/github/action-build-and-push/action.yml:1]: Very nice!

File: ci/github/action-build-and-push/print_as_header.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    print("| " + message + " " * spaces_length + " |")
+    print("-" * computed_length)
+
+def main():
+    input_message = " ".join(sys.argv[1:])
+    assemble_header(input_message)
+    
+
+if __name__ == '__main__':
+    main()
```
Comment (pcrespov) [inline ci/github/action-build-and-push/print_as_header.py:19]: MINOR: run black in all python files ...

File: ci/github/action-build-and-push/print_as_header.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+def assemble_header(message, default_length=80):
+    total_message_length = 4 + len(message)
+    computed_length = max(default_length, total_message_length)
+
+    spaces_length = default_length - len(message) - 4
+    print("-" * computed_length)
+    print("| " + message + " " * spaces_length + " |")
+    print("-" * computed_length)
+
+def main():
```
Comment (pcrespov) [inline ci/github/action-build-and-push/print_as_header.py:13]: MINOR: I also like having a separate free-function for main. This even allows testing separatly the main entrypoint... but in order to be totally coherent with this idea, I also expose the arguments of ``sys``, i.e.

```python
from typing import Optional, List, Any

def main(args: Optional[List[Any]]=None):
     if args is None:
         args = sys.argv[1:]

```

As I said, it does not change the behavior in this case but it is coherent with the idea of having a separate ``main`` for the main entrypoint

File: docker-registry-sync/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,21 @@
+# based on https://github.com/xelalexv/dregsy/blob/85b34eb8a0c7ffc1b7c610bc8de1388230a12551/Dockerfile
+FROM xelalex/dregsy@sha256:9b0fe252cb8bf88f67871efd6559ef89b76013c46dfbc49b961d396a4b8d1887
+
+LABEL maintainer "neagu@itis.swiss"
+
+RUN apk add python3
+RUN pip3 install --upgrade pip setuptools
```
Comment (pcrespov) [inline docker-registry-sync/Dockerfile:7]: MINOR: we also tend to upgrade ``wheels`` together with these two. 
MINOR: list every item in a different line (as above)

File: docker-registry-sync/Makefile [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	$(foreach iid,$(shell docker images */$(1):* -q | sort | uniq),\
+		docker image inspect $(iid) | jq '.[0] | .RepoTags, .ContainerConfig.Labels';)
+endef
+
+.PHONY: info
+info: ## gives some information about the docker image
+	@docker images */${IMAGE_NAME}:*;\
+	$(call show-meta,${IMAGE_NAME})
+
+# Used by GitHub Actions, these should be defined by each project
```
Comment (pcrespov) [inline docker-registry-sync/Makefile:100]: VERY NICE!!

File: docker-registry-sync/reposync/src/reposync/dregsy_config.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+from .prepare_stages import Stage, List
+from .utils import encode_credentials
+
+
+class BaseSerializable:
+    def to_dict(self) -> Dict:
+        raise NotImplementedError("Must implement in subclass")
+
+
+@dataclass
```
Comment (pcrespov) [inline docker-registry-sync/reposync/src/reposync/dregsy_config.py:16]: dataclass! fancy! :-)

File: docker-registry-sync/reposync/src/reposync/dregsy_config.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        return result
+
+
+@dataclass
+class TaskRegistry(BaseSerializable):
+    registry: str
+    auth: str
+    skip_tls_verify: bool
+
+    def to_dict(self) -> Dict:
```
Comment (pcrespov) [inline docker-registry-sync/reposync/src/reposync/dregsy_config.py:38]: Why don't  you use [asdict](https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict) here?? You can use a special factory (second argument) that convert ``_`` int ``-``.

File: docker-registry-sync/reposync/src/reposync/dregsy_config.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+            "name": self.name,
+            "verbose": self.verbose,
+            "source": self.source,
+            "target": self.target,
+            "mappings": self.mappings,
+        }
+
+
+@dataclass
+class DregsyYAML(BaseSerializable):
```
Comment (pcrespov) [inline docker-registry-sync/reposync/src/reposync/dregsy_config.py:65]: Dregsy = Docker Registry Sync? Are you following this project rationale? https://github.com/xelalexv/dregsy

File: docker-registry-sync/reposync/src/reposync/dregsy_config.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+    relay: str
+    skopeo: Dict[str, str]
+    tasks: List[Task]
+
+    def to_dict(self) -> Dict:
+        return {"relay": self.relay, "skopeo": self.skopeo, "tasks": self.tasks}
```
Comment (pcrespov) [inline docker-registry-sync/reposync/src/reposync/dregsy_config.py:71]: as above... asdict?

File: docker-registry-sync/reposync/src/reposync/dregsy_config.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+@dataclass
+class Task(BaseSerializable):
+    name: str
+    verbose: bool
+    source: TaskRegistry
+    target: TaskRegistry
+    mappings: List[Mapping]
+
+    def to_dict(self) -> Dict:
```
Comment (pcrespov) [inline docker-registry-sync/reposync/src/reposync/dregsy_config.py:54]: as above... asdict?

File: docker-registry-sync/reposync/src/reposync/dregsy_config.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+                    auth=source_auth,
+                    skip_tls_verify=stage.from_obj.source.skip_tls_verify,
+                ).to_dict(),
+                target=TaskRegistry(
+                    registry=to_obj.destination.url,
+                    auth=target_auth,
+                    skip_tls_verify=to_obj.destination.skip_tls_verify,
+                ).to_dict(),
+                mappings=[
+                    Mapping(
```
Comment (pcrespov) [inline docker-registry-sync/reposync/src/reposync/dregsy_config.py:107]: `Mapping(..).to_dict()`??  why not simply ``dict(a=b,...)`` ?

File: docker-registry-sync/reposync/src/reposync/main.py [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+        help="if an error occurs do not continue sync",
+    )
+    # add exit on first error to force a quit
+    args = parser.parse_args()
+    config_file = args.configfile
+
+    input_file = Path(config_file)
+
+    if not input_file.is_file():
+        print(f"Provided file '{input_file}' does not exist.")
```
Comment (pcrespov) [inline docker-registry-sync/reposync/src/reposync/main.py:106]: I think you can do this directly in ``argparse`` if you defined the argument as [filetype objects](https://docs.python.org/3/library/argparse.html#filetype-objects)? not 100% sure though ...

File: ci/github/action-build-and-push/print_as_header.py (hunk 4 above)
Comment (GitHK) [reply ci/github/action-build-and-push/print_as_header.py:13]: I like this suggestions. I will implement it for future reference.

**Pleas note:** I am leveraging the Python from the GitHub action's base image which is 3.4 and do not intend updating it. No typing support will be added here.

File: docker-registry-sync/reposync/src/reposync/dregsy_config.py (hunk 12 above)
Comment (GitHK) [reply docker-registry-sync/reposync/src/reposync/dregsy_config.py:107]: I needed extra functionality and the name is just unfortunate, as I mimicked the configuration file's YAML key name.

File: docker-registry-sync/reposync/src/reposync/main.py (hunk 13 above)
Comment (GitHK) [reply docker-registry-sync/reposync/src/reposync/main.py:106]: worked, thanks 👍

File: docker-registry-sync/reposync/src/reposync/dregsy_config.py (hunk 9 above)
Comment (GitHK) [reply docker-registry-sync/reposync/src/reposync/dregsy_config.py:65]: Yes, this project runs on top of it.

File: docker-registry-sync/reposync/src/reposync/dregsy_config.py (hunk 8 above)
Comment (pcrespov) [reply docker-registry-sync/reposync/src/reposync/dregsy_config.py]: Did you check this?

File: docker-registry-sync/reposync/src/reposync/utils.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
...
+
+
+@contextmanager
+def temp_configuration_file(stage_name: str, yaml_string: str) -> IO[Any]:
+    target_file = Path("/tmp") / f"{sanitize(stage_name.replace(' ', ''))}.yaml"
+    try:
+        f = target_file.open("w")
+        yield f
+    except Exception:
+        import traceback
```
Comment (pcrespov) [inline docker-registry-sync/reposync/src/reposync/utils.py:36]: Why traceback is inside?

File: docker-registry-sync/reposync/src/reposync/main.py [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+
+def load_yaml_from_file(input_file: Path) -> str:
+    with input_file.open("r") as f:
+        return yaml.safe_load(f)
+
+
+def error_exit(start_date):
+    print("\nThere were errors during sync, check the logs above")
+    print(f"Error after after: {datetime.datetime.utcnow() - start_date}\n")
```
[the remark this answers is not in the document]
Comment (GitHK) [reply docker-registry-sync/reposync/src/reposync/main.py:21]: No, the wrapped tool already has a retry. If an error is detected it gives up.

File: docker-registry-sync/reposync/src/reposync/utils.py (hunk 14 above)
Comment (GitHK) [reply docker-registry-sync/reposync/src/reposync/utils.py:36]: not a good reason, moved outside.

File: docker-registry-sync/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+COPY reposync /app/reposync
+
+WORKDIR /app/reposync
+
+RUN pip3 install .
+
+# remove install directory
+RUN rm -rf /app
+
+CMD ["run-reposync", "-c", "/etc/cfg.yaml"]
```
Comment (sanderegg) [inline docker-registry-sync/Dockerfile:23]: why not using ENTRYPOINT instead?

File: docker-registry-sync/Makefile [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+
+.PHONY: github-ci-build
+github-ci-build: ## reserved for Git Hub CI to build image
+	@make build
+	# checkk the version for stable
+ifeq (${DOCKER_REG_SYNC_VERSION}, ${DOCKER_REG_SYNC_TAS_AS_STABLEVERSION})
+	docker tag ${DOCKER_REGISTRY}/${IMAGE_NAME}:${DOCKER_REG_SYNC_VERSION} ${DOCKER_REGISTRY}/${IMAGE_NAME}:stable
+endif
+	@make info
+
```
Comment (sanderegg) [inline docker-registry-sync/Makefile:116]: so I am not sure now I get it correctly, does the built service need to provide ```make github-ci-build``` or ```make build``` is enough?

File: docker-registry-sync/reposync/setup.py [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+test_requirements = read_reqs(current_dir / "requirements" / "test.txt")
+
+setup(
+    name="reposync",
+    version="0.1.0",
+    packages=find_packages(where="src"),
+    package_dir={"": "src",},
+    include_package_data=True,
+    entry_points={"console_scripts": ["run-reposync=reposync.main:main",]},
+    python_requires=">=3.7",
```
Comment (sanderegg) [inline docker-registry-sync/reposync/setup.py:26]: thus the dataclass stuff. nice!

File: docker-registry-sync/reposync/src/reposync/utils.py [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+import json
+import os
+
+from typing import IO, Any
+from contextlib import contextmanager
+from pathlib import Path
+from sanitize_filename import sanitize
+
+
+def encode_string(message, encoding="utf-8"):
```
Comment (sanderegg) [inline docker-registry-sync/reposync/src/reposync/utils.py:12]:
```suggestion
def encode_string(message: str, encoding: str="utf-8") -> bytes:
```

File: docker-registry-sync/reposync/src/reposync/utils.py [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+    return base64.b64encode(message.encode(encoding)).decode(encoding)
+
+
+def encode_credentials(username, password):
```
Comment (sanderegg) [inline docker-registry-sync/reposync/src/reposync/utils.py:16]:
```suggestion
def encode_credentials(username: str, password: str) -> bytes:
```

File: ci/github/action-build-and-push/README.md [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+- `REGISTRY_USER`[**mandatory**] registry username
+- `REGISTRY_PASSWORD`[**mandatory**] registry password or access token
+- `TARGET_PROJECT_PATH`[**mandatory**] registry password or access token
+
+Inside the `TARGET_PROJECT_PATH` the action expects to find a Makefile as the following commands will be issued:
+
+- `make github-ci-pull` will attempt to pull a previous image, this command may fail and the action will continue
+- `make github-ci-build` builds the image
+- `make github-ci-tests` runs the tests packaged with the image
+- `make github-ci-push` pushes the image to the registry
```
[the remark this answers is not in the document]
Comment (GitHK) [reply ci/github/action-build-and-push/README.md:19]: As explained this is meant to be used in a generic way. These stages were added for convenience and to embrace loose coupling between components.

File: ci/github/action-build-and-push/entrypoint.sh [hunk 22]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
+
+set -e
```
[the remark this answers is not in the document]
Comment (GitHK) [reply ci/github/action-build-and-push/entrypoint.sh:3]: thanks, I was searching for this and forgot the name, glad you piked it up.

File: docker-registry-sync/Makefile (hunk 17 above)
Comment (GitHK) [reply docker-registry-sync/Makefile:116]: explained above

**Pull request merged**